providerquerymanager/providerquerymanager.go

```diff
@@ -0,0 +1,407 @@
+		network:                      network,
+		providerQueryMessages:        make(chan providerQueryMessage, 16),
+		providerRequestsProcessing:   make(chan *findProviderRequest),
+		incomingFindProviderRequests: make(chan *findProviderRequest),
+		inProgressRequestStatuses:    make(map[cid.Cid]*inProgressRequestStatus),
+		findProviderTimeout:          defaultTimeout,
+	}
+}
+
+// Startup starts processing for the ProviderQueryManager.
+func (pqm *ProviderQueryManager) Startup() {
+	go pqm.run()
+}
+
+type inProgressRequest struct {
+	providersSoFar []peer.ID
+	incoming       chan peer.ID
+}
+
+// SetFindProviderTimeout changes the timeout for finding providers
+func (pqm *ProviderQueryManager) SetFindProviderTimeout(findProviderTimeout time.Duration) {
+	pqm.timeoutMutex.Lock()
+	pqm.findProviderTimeout = findProviderTimeout
+	pqm.timeoutMutex.Unlock()
+}
+
+// FindProvidersAsync finds providers for the given block.
+func (pqm *ProviderQueryManager) FindProvidersAsync(sessionCtx context.Context, k cid.Cid) <-chan peer.ID {
+	inProgressRequestChan := make(chan inProgressRequest)
+
+	select {
+	case pqm.providerQueryMessages <- &newProvideQueryMessage{
+		k:                     k,
+		inProgressRequestChan: inProgressRequestChan,
+	}:
+	case <-pqm.ctx.Done():
+		return nil
+	case <-sessionCtx.Done():
+		return nil
+	}
+
+	var receivedInProgressRequest inProgressRequest
+	select {
+	case <-pqm.ctx.Done():
+		return nil
+	case <-sessionCtx.Done():
+		return nil
+	case receivedInProgressRequest = <-inProgressRequestChan:
+	}
+
+	return pqm.receiveProviders(sessionCtx, k, receivedInProgressRequest)
+}
+
+func (pqm *ProviderQueryManager) receiveProviders(sessionCtx context.Context, k cid.Cid, receivedInProgressRequest inProgressRequest) <-chan peer.ID {
+	// maintains an unbuffered queue for incoming providers for given request for a given session
+	// essentially, as a provider comes in, for a given CID, we want to immediately broadcast to all
+	// sessions that queried that CID, without worrying about whether the client code is actually
+	// reading from the returned channel -- so that the broadcast never blocks
+	// based on: https://medium.com/capital-one-tech/building-an-unbounded-channel-in-go-789e175cd2cd
+	returnedProviders := make(chan peer.ID)
+	receivedProviders := append([]peer.ID(nil), receivedInProgressRequest.providersSoFar[0:]...)
+	incomingProviders := receivedInProgressRequest.incoming
+
+	go func() {
+		defer close(returnedProviders)
+		outgoingProviders := func() chan<- peer.ID {
+			if len(receivedProviders) == 0 {
+				return nil
+			}
+			return returnedProviders
+		}
+		nextProvider := func() peer.ID {
+			if len(receivedProviders) == 0 {
+				return ""
+			}
+			return receivedProviders[0]
+		}
+		for len(receivedProviders) > 0 || incomingProviders != nil {
+			select {
+			case <-pqm.ctx.Done():
+				return
+			case <-sessionCtx.Done():
+				pqm.providerQueryMessages <- &cancelRequestMessage{
```

Comment:
Probably needs to select on `pqm.ctx.Done()`.

Reply:
Also, I think this needs to read from `incomingProviders` at the same time. Otherwise, we're relying on the buffer in `providerQueryMessages` being large enough to fit this message.

Reply:
right. I hate go. :P will fix.

Reply:
fixed, I think